### ISSUE TITLE
[Bug: Navigation timeout 30000 ms exceeded

### ISSUE THREAD

**Issue by rakshithrai77** [puppeteer/puppeteer#12077]
### Minimal, reproducible example

```TypeScript
const browser = await puppeteer.launch({
			headless: true,
			args: ["--no-sandbox", "--font-render-hinting=none"],
		});
		const page = await browser.newPage();

		await page.setContent(html ,{waitUntil:"domcontentloaded"});
		const pdfBuffer = await page.pdf(options);

		// Close browser
		await browser.close();
```

### Error string

Navigation timeout 30000 ms exceeded

### Bug behavior

- [ ] Flaky
- [X] PDF

### Background

Getting Navigation timeout error 30000 ms exceeded when i deploy code to server
When I removed external resources from html i am able to generate the pdf.
But if HTML has external scripts this error throws up in server Navigation timeout error 30000 ms,

### Expectation

PDF should be generated with no issues

### Reality

Getting Navigation timeout error 30000 ms exceeded when i deploy code to server
When I removed external resources from html i am able to generate the pdf.
But if HTML has external scripts this error throws up in server Navigation timeout error 30000 ms,

### Puppeteer version

21.9.0

### Node version

16.20.0

### Package manager

npm

### Package manager version

8.19.4

### Operating system

Linux

**Comment by Lightning00Blade** [comment]
Does the issue reproduce for you locally? If so could you either provide a fully reproducible example with variables included.
Or if its only after deploying collect logs via setting `DEBUG=puppeteer:*` ([Debugging](https://gist.github.com/)) and share them here or via a [GitHub Gist](https://gist.github.com/).
Note if you use any sensitive data with puppeteer that may show in the logs, so make sure to remove it before posting.

**Comment by OrKoN** [comment]
Consider using the `dumpio: true` flag when launching Puppeteer on the server to make sure the browser does not crash and has all required dependencies. Additionally, make sure the server has sufficient memory and cpu resources to run the browser or increase the navigation timeout. If your HTML content references subresources make sure they are accessible from the server.

**Comment by Daijobou** [comment]
I have the same error message, but in another case. I load a page and want work with it. The error comes random. In a previous version 19.11.1 the error message was:

> (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\common\LifecycleWatcher.js:174
>     return new Errors_js_1.TimeoutError(errorMessage);
>            ^
> 
> TimeoutError: Navigation timeout of 30000 ms exceeded
>     at LifecycleWatcher._LifecycleWatcher_createTimeoutPromise (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\common\LifecycleWatcher.js:174:12)

in latest version 22.6.0 of puppeteer its now:

> C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\util\Deferred.js:59
>             this.#timeoutError = new Errors_js_1.TimeoutError(opts.message);
>                                  ^
> 
> TimeoutError: Navigation timeout of 30000 ms exceeded
>     at new Deferred (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\util\Deferred.js:59:34)
>     at Deferred.create (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\util\Deferred.js:21:16)
>     at new LifecycleWatcher (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\cdp\LifecycleWatcher.js:65:60)
>     at CdpFrame.goto (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\cdp\Frame.js:136:29)
>     at CdpFrame.<anonymous> (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\util\decorators.js:98:27)
>     at CdpPage.goto (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\api\Page.js:590:43)
>     at scrape (C:\Users\username\myscript.js:27:16)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async C:\Users\username\myscript.js:76:35

**Comment by pontiggia** [comment]
> I have the same error message, but in another case. I load a page and want work with it. The error comes random. In a previous version 19.11.1 the error message was:
> 
> > (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\common\LifecycleWatcher.js:174
> > return new Errors_js_1.TimeoutError(errorMessage);
> > ^
> > TimeoutError: Navigation timeout of 30000 ms exceeded
> > at LifecycleWatcher._LifecycleWatcher_createTimeoutPromise (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\common\LifecycleWatcher.js:174:12)
> 
> in latest version 22.6.0 of puppeteer its now:
> 
> > C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\util\Deferred.js:59
> > this.#timeoutError = new Errors_js_1.TimeoutError(opts.message);
> > ^
> > TimeoutError: Navigation timeout of 30000 ms exceeded
> > at new Deferred (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\util\Deferred.js:59:34)
> > at Deferred.create (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\util\Deferred.js:21:16)
> > at new LifecycleWatcher (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\cdp\LifecycleWatcher.js:65:60)
> > at CdpFrame.goto (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\cdp\Frame.js:136:29)
> > at CdpFrame. (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\util\decorators.js:98:27)
> > at CdpPage.goto (C:\Users\username\node_modules\puppeteer-core\lib\cjs\puppeteer\api\Page.js:590:43)
> > at scrape (C:\Users\username\myscript.js:27:16)
> > at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> > at async C:\Users\username\myscript.js:76:35

Any solutions? I have the same error when deploy in server:
Logging in...

/usr/src/app/node_modules/puppeteer-core/lib/cjs/puppeteer/util/Deferred.js:59

this.#timeoutError = new Errors_js_1.TimeoutError(opts.message);

^

TimeoutError: Navigation timeout of 30000 ms exceeded

at new Deferred (/usr/src/app/node_modules/puppeteer-core/lib/cjs/puppeteer/util/Deferred.js:59:34)

at Deferred.create (/usr/src/app/node_modules/puppeteer-core/lib/cjs/puppeteer/util/Deferred.js:21:16)

at new LifecycleWatcher (/usr/src/app/node_modules/puppeteer-core/lib/cjs/puppeteer/cdp/LifecycleWatcher.js:66:60)

at CdpFrame.goto (/usr/src/app/node_modules/puppeteer-core/lib/cjs/puppeteer/cdp/Frame.js:136:29)

**Comment by xerath** [comment]
Same error.. when deploying to AWS Apprunner (containers)

Randomly it works..  literally, I copied/pasted the code from where it wasn't working to a healthcheck endpoint where it suddenly worked..  deffo seems to be some randomness/intermittency to it when running in the cloud for some reason

**Comment by xerath** [comment]
Just to add some stuff..  

1) I used http module to hit the same site before pupeteer to ensure DNS working and a valid internet route
2) When I try/catch puppeteer page.gotos..  and allow the process to run..  it will literally fail 2/3 of the time with 'navigation timeout 30000 ms exceeded' but then succeed some of the time too.

Totally intermittent issue that only happens on containerised (headless: true) setups in the cloud.

**Comment by John251314** [comment]
Yes, I have the same issues, when run on ubuntu server; puppeteer seem can not connect to the Internet

**Comment by John251314** [comment]
> Yes, I have the same issues, when run on ubuntu server; puppeteer seem can not connect to the Internet

It`s foo, my server is too slow, modify timeout to 60s, then it`s ok!

**Comment by ivanalemunioz** [comment]
If any of you are wasting days debugging Puppeteer crashes, I'm creating a new service to make it easier: [Buglesstack](https://buglesstack.com/). It's currently in beta, so I'd be glad if any of you tried it out!

**Comment by branstetterz** [comment]
We're encountering a TimeoutError: Navigation timeout of 30000 ms exceeded when running our Puppeteer script (feedCamels.js) in GitHub Actions. The workflow fails at the page navigation step despite working locally.

**Comment by ivanalemunioz** [comment]
Hi, everyone. I’ve been working on a solution to visually debug my browser automation crashes, and I’ve now made it free for the public to use: https://buglesstack.com

I don’t mean to spam or bother you; I just want to share this solution. If you have any questions or suggestions for improvement, feel free to contact me.

Thanks for reading! ❤️

**Comment by PedroContipelli** [comment]
Was running into a similar issue when trying to deploy my service, and it turned out to be a memory issue. 512 MB was not enough to properly run chromium. Scaling up to 2GB fixed it. Also running with tons of flags to keep it lean.
```js
browser = await puppeteer.launch({
  headless: true,
  args: [
    "--no-sandbox",
    "--disable-setuid-sandbox",
    "--disable-dev-shm-usage",
    "--disable-extensions",
    "--disable-background-networking",
    "--disable-default-apps",
    "--disable-sync",
    "--disable-translate",
    "--single-process",
    "--no-zygote",
    "--disable-gpu"
  ]
});
```

**Comment by npn6094-alt** [comment]
hey hai  still i get 
TimeoutError: Navigation timeout of 30000 ms exceeded
    at new Deferred (file:///home/ubuntu/captcha-OCT/node_modules/puppeteer-core/lib/esm/puppeteer/util/Deferred.js:57:34)
    at Deferred.create (file:///home/ubuntu/captcha-OCT/node_modules/puppeteer-core/lib/esm/puppeteer/util/Deferred.js:18:16)
    at new LifecycleWatcher (file:///home/ubuntu/captcha-OCT/node_modules/puppeteer-core/lib/esm/puppeteer/cdp/LifecycleWatcher.js:70:46)
    at CdpFrame.goto (file:///home/ubuntu/captcha-OCT/node_modules/puppeteer-core/lib/esm/puppeteer/cdp/Frame.js:145:29)
    at CdpFrame.<anonymous> (file:///home/ubuntu/captcha-OCT/node_modules/puppeteer-core/lib/esm/puppeteer/util/decorators.js:101:27)
    at CdpPage.goto (file:///home/ubuntu/captcha-OCT/node_modules/puppeteer-core/lib/esm/puppeteer/api/Page.js:568:43)
    at runCloudflareWAFTest (file:///home/ubuntu/captcha-OCT/jan.js:21:14)



** need to solve without third party api key or token** 

code :

import puppeteer from 'puppeteer';
import Xvfb from 'xvfb';

const realBrowserOption = {
  args: ['--start-maximized'],
  turnstile: true,
  headless: false,
  customConfig: {},
  connectOption: {
    defaultViewport: null,
  },
  plugins: [],
};

const runCloudflareWAFTest = async () => {
  const { page, browser } = await connect(realBrowserOption);

  await page.goto("https://www.google.com/search?q=stealth");
  await new Promise((r) => setTimeout(r, 5000));
  await page.screenshot({ path: `screenshot-2.png`, fullPage: true });

  await new Promise((r) => setTimeout(r, 5000));
  await browser.close();
};

runCloudflareWAFTest();


// ---------- Connect Function ----------
async function connect({
  args = [],
  headless = 'new',
  customConfig = {},
  proxy = {},
  turnstile = false,
  connectOption = {},
  disableXvfb = false,
  plugins = [],
  ignoreAllFlags = false,
} = {}) {
  let xvfbsession = null;

  // Start XVFB for Linux
 // if (process.platform === "linux" && !disableXvfb) {
    try {
      xvfbsession = new Xvfb({
        silent: true,
        xvfb_args: ["-screen", "0", "1920x1080x24", "-ac"],
      });
      xvfbsession.startSync();
    } catch (err) {
      console.warn("Xvfb not found. Please install it: sudo apt-get install xvfb\n" + err.message);
    //}
  }

  const launchArgs = [
    ...args,
    "--no-sandbox",
    "--disable-dev-shm-usage",
    "--disable-blink-features=AutomationControlled",
    "--start-maximized",
    ...(proxy?.host && proxy?.port ? [`--proxy-server=${proxy.host}:${proxy.port}`] : []),
    "--disable-dev-shm-usage",
    "--disable-extensions",
    "--disable-background-networking",
    "--disable-default-apps",
    "--disable-sync",
    "--disable-translate",
    "--single-process",
    "--no-zygote",
    "--disable-gpu"
  ];

  const browser = await puppeteer.launch({
    headless,
    args: launchArgs,
    defaultViewport: null,
    ...customConfig,
  });

  const page = await browser.newPage();
  return { browser, page };
}



some one give solution
